modules/concurrency/src/spinner.cpp

```diff
@@ -64,34 +56,76 @@ void Spinner::spin() {
 
   try {
     start_timestamp_ = std::chrono::system_clock::now();
+    uint64_t spin_count_{ 0 };
 
     while (!stop_requested_.load()) {
-      const auto spin_result = stoppable_callback_();
+      // State machine to handle the spinner lifecycle.
```

Comment:
here the code is a bit too convoluted and hard to follow, I would try to first break it down into steps:
(I do not even think you need the states in the enum).
```c++
void Spinner::spin() {
  // This is the main loop of the whole state machine
  while (!stop_requested_)) {
    auto res = initialize();
    if (!res) { break; } // This means we have trying initializing for a while and we couldn't so there is nothing more we can do
    auto res = execute();
    if (shall_stop_cb(res)) break; // We restart from init
    else continue;
  }
  termination_cb();
}

auto Spinner::initialize() -> bool {
  while (!stop_requested_) {
     try {
       init_cb()
       return true;
     } catch() {
        if (should_re_init()) continue;
        else return false;
     }
  }
}

auto Spinner::execute() -> bool {
  try {
     while (!stop_requested_)) {
        spin_cb()
         std::this_thread::sleep_for(...);
     }
  } catch() {
     return false;
  }
}

```

Reply:
This is rather a conceptual question. Your solution removes observability from the states, and it has some implications
* all transitions are always executed, unless we add additional checks
* how would you implement these checks? e.g. is_initialized? I assume you would add a state
* the proposed solution works as long as the logic chain is linear, but does not allow for branching, see. e.g. the missing branching surrounding `stop?`. I see that we can still fix it for the current logic, but extendability in the future will suffer.
* Extendability: we extended `Spinner` to `StoppableSpinner`. Now we extend to `ReInitableAndStoppableSpinner`. There's a good chance we will add additional capabilities in the future. I thus prefer an extendable design.

Reply:
If you are in `execute` you can assume that you have successfully initialized.
I think what confuses me is that in your solution you have a single loop that handles all the states. I am not a state machine expert so maybe that is the correct way to go, but it make also the code much more complex. 
I was proposing a more hierarchical solution. Where you have a top level loop and then inside you run the different states.
I am not sure I understand the missing branch surrounding stop. In my proposal you check if you want to start from scratch or not. 

In my opinion, we either take the time and implement a proper nice state machine abstraction, maybe even check at some 3rd party for reference, or it is fine to avoid early generalization and overfit on our problem at end.

Reply:
Can I close this one @filippobrizzi ?

(The proposed code is very elegant by the way, just wanted to point that out)